frame/support/src/traits/hooks.rs

```diff
@@ -359,10 +359,64 @@ pub trait OnTimestampSet<Moment> {
 #[cfg(test)]
 mod tests {
 	use super::*;
+	use sp_io::TestExternalities;
+
+	#[test]
+	fn on_runtime_upgrade_pre_post_executed_tuple() {
+		crate::parameter_types! {
+			pub static Pre: Vec<&'static str> = Default::default();
+			pub static Post: Vec<&'static str> = Default::default();
+		}
+
+		macro_rules! impl_test_type {
+			($name:ident) => {
+				struct $name;
+				impl OnRuntimeUpgrade for $name {
+					fn on_runtime_upgrade() -> Weight {
+						Default::default()
+					}
+
+					#[cfg(feature = "try-runtime")]
+					fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+						Pre::mutate(|s| s.push(stringify!($name)));
+						Ok(Vec::new())
+					}
+
+					#[cfg(feature = "try-runtime")]
+					fn post_upgrade(_: Vec<u8>) -> Result<(), &'static str> {
+						Post::mutate(|s| s.push(stringify!($name)));
+						Ok(())
+					}
+				}
+			};
+		}
+
+		impl_test_type!(Foo);
+		impl_test_type!(Bar);
+		impl_test_type!(Baz);
+
+		TestExternalities::default().execute_with(|| {
+			Foo::try_on_runtime_upgrade(true).unwrap();
+			// todo unify the API for storage parameter_types! output.
```

Comment:
I seem to have had a leftover TODO here which you can also remove and optionally pursue in a separate PR. 

Here's what is going on. 

The `parameter_types` works in 3 modes. 

1. static (generates thread_local stuff)
2. storage 
3. const 

All of which generate a `Get` implementation, backed by something, along some other helper functions like `get`, `set`, `mutate` and `take` (kinda similar to storage items..). 

It seemed to me that the set of helper functions that is generated by `parameter_types!` is not uniform across different modes. 

If interested, you can convert this comment into an issue and self-assign.

Reply:
Sounds good, created https://github.com/paritytech/substrate/issues/13795